### PR TITLE
Gossip PING/PONG control messages

### PR DESCRIPTION
**Pull request by Nashatyrev**
Internal draft PR 

- `min_ping_period` seems too restrictive. `max_ping_rate` [per minute] seems more flexible

### CODE REVIEW

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+## Parameters
+
+This section lists the configuration parameters that needs to agreed on across clients to avoid
+peer penalizations
+
+| Parameter         | Description                                                            | Reasonable Default |
+|-------------------|------------------------------------------------------------------------|--------------------|
+| `min_ping_period` | minimal delay between sending next `PING` and previous received `PONG` | 1 second (?)       |
+| `max_ping_rate`   | (better alternative) maximum number of `PING` messages per minute      | 60                 |
```
Comment (Menduist) [inline pubsub/gossipsub/gossipsub-v1.2.md]: max_ping_rate does seem better, but 60/min is far too low
If we ping every other RTT with 20ms ping, we are already at 25ping/s

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * [either `min_ping_period` or `max_ping_rate`]
+
+## PING/PONG Messages
+
+### Sender
+
+Sender MAY send `PING` message (either standalone or piggy tailing) at any time. Sender is free choose 
+any `id` however it might be reasonable to select unique numbers across the session.
```
Comment (Menduist) [inline pubsub/gossipsub/gossipsub-v1.2.md:40]: Maybe unpredictable by the other peer? Avoids weird attacks I guess

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+Upon receiving a `PING` message the responder
+- MUST respond with a `PONG` message with the same `id` value
+- SHOULD check if the `min_ping_period` is not violated (otherwise the peer MAY be penalized)
+  [or alternatively `max_ping_rate` is not voilated]
+- SHOULD make his best to respond back with a corresponding `PONG` as soon as possible. That means:
+  - SHOULD NOT queue the message for later send (e.g. on heartbeat or piggy tailing for a next publish message)
+  - if there are messages queued for sending to the peer that queue SHOULD be disregarded and the 
+    the `PONG` message should be flushed immediately
+
+### Usage scenarios
```
Comment (Menduist) [inline pubsub/gossipsub/gossipsub-v1.2.md:61]: Can also talk about the connection "keep alive" (or keep warm I guess)

File: pubsub/gossipsub/gossipsub-v1.2.md (hunk 1 above)
Comment (Nashatyrev) [reply pubsub/gossipsub/gossipsub-v1.2.md]: Ok will leave `max_ping_rate `
>  but 60/min is far too low

Yep, but this a 'reasonable default' for normal Libp2p usage 😆 
We will adjust this option for Eth2 when we need

File: pubsub/gossipsub/gossipsub-v1.2.md (hunk 2 above)
Comment (Nashatyrev) [reply pubsub/gossipsub/gossipsub-v1.2.md:40]: Do you think there could be any security implications?

File: pubsub/gossipsub/gossipsub-v1.2.md (hunk 3 above)
Comment (Nashatyrev) [reply pubsub/gossipsub/gossipsub-v1.2.md:61]: Oh, that's actually too weird use case from my perspective. 
I'm not sure that dirty hack is worth mentioning in libp2p spec 😆

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
+# Gossipsub v1.2
+
+# Overview
+
+This document aims to provide a minimal extension to the [gossipsub
+v1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md)
+protocol.
+
+The proposed extensions are backwards-compatible and aim to add utility control messages for 
+better network efficiency (delivery latency and total throughput) tuning of gossip protocol
```
Comment (mkalinin) [inline pubsub/gossipsub/gossipsub-v1.2.md]:
```suggestion
better network efficiency (delivery latency and total throughput) tuning gossip protocol 
```

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+| Parameter         | Description                                                            | Reasonable Default |
+|-------------------|------------------------------------------------------------------------|--------------------|
+| `max_ping_rate`   | maximum number of `PING` messages per minute      | 60                 |
+
+## PING/PONG Messages
+
+### Sender
+
+Sender MAY send `PING` message (either standalone or piggy tailing) at any time. Sender is free choose
```
Comment (mkalinin) [inline pubsub/gossipsub/gossipsub-v1.2.md]:
```suggestion
Sender MAY send `PING` message (either standalone or piggy tailing) at any time. Sender is free to choose 
```

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+Sender MAY send `PING` message (either standalone or piggy tailing) at any time. Sender is free choose 
+any `id` however it might be reasonable to select unique numbers across the session. 
+
+Sender MUST NOT send `PING` message to a peer if `max_ping_rate` `PING` message were already sent 
+to that peer during the last minute.
+
+An implementation is free to choose a reasonable timeout (e.g. 60 seconds) for waiting for a corresponding 
+`PONG` message. However it is worth considering that a `PONG` message might be 
+[... blocked by another message which is already on the wire] and not setting the timeout too small.
```
Comment (mkalinin) [inline pubsub/gossipsub/gossipsub-v1.2.md]:
```suggestion
[... blocked by another message which is already on the wire, thus the timeout should not be too small.
```

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+If a `PONG` timeout occurs that SHOULD be treated as protocol malfunction and the peer connection
```
Comment (mkalinin) [inline pubsub/gossipsub/gossipsub-v1.2.md]:
```suggestion
`PONG` timeout exceeding SHOULD be treated as protocol malfunction and the peer connection 
```

Comment (mkalinin) [reply pubsub/gossipsub/gossipsub-v1.2.md]: A question: if any other message is received from a remote peer, should it reset the timeout? For instance, if a channel was busy by transmitting a big message from a slow peer a `PONG` might be too late and cause disconnection by timeout. I think defining a recommended timeout value in the spec should be useful

File: pubsub/gossipsub/gossipsub-v1.2.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+- SHOULD make his best to respond back with a corresponding `PONG` as soon as possible. That means:
+  - SHOULD NOT queue the message for later send (e.g. on heartbeat or piggy tailing for a next publish message)
+  - if there are messages queued for sending to the peer that queue SHOULD be disregarded and the 
+    the `PONG` message should be flushed immediately
+
+### Usage scenarios
+
+#### Latency estimation
+
+This is the most obvious and basic scenario of Ping/Pong usage. Gossip implementation may use the estimated
```
Comment (mkalinin) [inline pubsub/gossipsub/gossipsub-v1.2.md]:
```suggestion
This is the most obvious and basic scenario of Ping/Pong usage. Gossip implementation may use estimated 
```

**Pull request closed without merging**